File: iOS/Daily/Views/Habits/HabitCircleBreakdownTicker.swift
import SwiftUI
import DailyCore

/// A sleek, compact horizontal ticker capsule designed to fit comfortably inside
/// circular progress gauges (Bubbles and Smokes) to display breakdown totals per liquid or smoke type.
public struct HabitCircleBreakdownTicker: View {
    public let items: [HabitDrinkBreakdown]
    public let habitType: HabitType
    
    @State private var autoScrollOffset: CGFloat = 0
    
    public init(items: [HabitDrinkBreakdown], habitType: HabitType = .water) {
        self.items = items
        self.habitType = habitType
    }
    
    /// Aggregate items by primary beverage / tobacco category for clean, non-fragmented totals
    public var aggregatedItems: [AggregatedItem] {
        var map: [String: (amount: Double, unit: String, hexColor: String, iconName: String)] = [:]
        
        for item in items {
            let norm: (category: String, color: String, icon: String)
            if habitType == .water {
                norm = normalizeDrink(name: item.drink, defaultColor: item.hexColor, defaultIcon: item.iconName)
            } else {
                norm = normalizeSmoke(name: item.drink, defaultColor: item.hexColor, defaultIcon: item.iconName)
            }
            
            if let existing = map[norm.category] {
                map[norm.category] = (existing.amount + item.amount, existing.unit, existing.hexColor, existing.iconName)
            } else {
                map[norm.category] = (item.amount, item.unit, norm.color, norm.icon)
            }
        }
        
        return map.map { key, val in
            AggregatedItem(
                id: key,
                name: key,
                amount: val.amount,
                unit: val.unit,
                hexColor: val.hexColor,
                iconName: val.iconName
            )
        }.sorted { $0.amount > $1.amount }
    }
    
    public struct AggregatedItem: Identifiable {
        public let id: String
        public let name: String
        public let amount: Double
        public let unit: String
        public let hexColor: String
        public let iconName: String
    }
    
    private func normalizeDrink(name: String, defaultColor: String, defaultIcon: String) -> (category: String, color: String, icon: String) {
        let lower = name.lowercased()
        if lower.contains("coffee") || lower.contains("espresso") || lower.contains("cappuccino") || lower.contains("latte") {
            return ("Coffee", "#F59E0B", "cup.and.saucer.fill")
        } else if lower.contains("tea") || lower.contains("matcha") || lower.contains("infusion") {
            return ("Tea", "#84CC16", "mug.fill")
        } else if lower.contains("water") || lower.contains("glass") || lower.contains("bottle") {
            return ("Water", "#00E5FF", "drop.fill")
        } else {
            return (name, defaultColor, defaultIcon)
        }
    }
    
    private func normalizeSmoke(name: String, defaultColor: String, defaultIcon: String) -> (category: String, color: String, icon: String) {
        let lower = name.lowercased()
        if lower.contains("cigarette") || lower == "cig" || lower.contains("standard") {
            return ("Cigarette", "#EF4444", "flame.fill")
        } else if lower.contains("cigarillo") || (lower.contains("cigar") && !lower.contains("cigarette")) {
            return ("Cigarillo", "#A855F7", "flame")
        } else if lower.contains("heat") || lower.contains("iqos") || lower.contains("glo") || lower.contains("vape") {
            return ("Heated", "#3B82F6", "bolt.fill")
        } else if lower.contains("roll") {
            return ("Rolled", "#F97316", "leaf.fill")
        } else {
            return ("Cigarette", "#EF4444", "flame.fill")
        }
    }
    
    private func itemBadgeText(item: AggregatedItem, isSingle: Bool) -> String {
        let formattedAmount = item.unit == "ml" ? "\(Int(item.amount)) ml" : "\(Int(item.amount))"
        if isSingle {
            return "\(item.name): \(formattedAmount)"
        }
        
        if habitType == .water {
            if item.name == "Water" {
                return "\(formattedAmount)"
            } else {
                return "\(item.name) \(formattedAmount)"
            }
        } else {
            switch item.name {
            case "Cigarette": return "\(Int(item.amount)) Cig"
            case "Heated": return "\(Int(item.amount)) Heat"
            case "Rolled": return "\(Int(item.amount)) Roll"
            case "Cigarillo": return "\(Int(item.amount)) Cigar"
            default: return "\(Int(item.amount)) \(item.name)"
            }
        }
    }
    
    public var body: some View {
        let activeItems = aggregatedItems
        if !activeItems.isEmpty {
            ScrollViewReader { proxy in
                ScrollView(.horizontal, showsIndicators: false) {
                    HStack(spacing: 5) {
                        ForEach(Array(activeItems.enumerated()), id: \.element.id) { index, item in
                            HStack(spacing: 3) {
                                Image(systemName: item.iconName)
                                    .font(.system(size: 8.5, weight: .bold))
                                    .foregroundColor(Color(hex: item.hexColor))
                                
                                Text(itemBadgeText(item: item, isSingle: activeItems.count == 1))
                                    .font(.system(size: 9.5, weight: .bold, design: .rounded))
                                    .foregroundColor(.white.opacity(0.92))
                            }
                            .padding(.horizontal, 6)
                            .padding(.vertical, 2.5)
                            .background(
                                Color(hex: item.hexColor).opacity(0.22)
                            )
                            .clipShape(Capsule())
                            .overlay {
                                Capsule()
                                    .strokeBorder(Color(hex: item.hexColor).opacity(0.4), lineWidth: 0.8)
                            }
                            .id(index)
                        }
                    }
                    .padding(.horizontal, 6)
                    .padding(.vertical, 2)
                }
                .frame(maxWidth: 175)
                .frame(height: 22)
                .background(Color.black.opacity(0.40))
                .background(.ultraThinMaterial)
                .clipShape(Capsule())
                .overlay {
                    Capsule().strokeBorder(Color.white.opacity(0.18), lineWidth: 0.8)
                }
                .shadow(color: Color.black.opacity(0.35), radius: 3, x: 0, y: 1)
            }
        }
    }
}

File: Android/app/src/main/java/com/intellidream/daily/presentation/habits/HabitCircleBreakdownTicker.kt
package com.intellidream.daily.presentation.habits

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.Bolt
import androidx.compose.material.icons.rounded.Coffee
import androidx.compose.material.icons.rounded.Eco
import androidx.compose.material.icons.rounded.EmojiFoodBeverage
import androidx.compose.material.icons.rounded.LocalFireDepartment
import androidx.compose.material.icons.rounded.SmokingRooms
import androidx.compose.material.icons.rounded.WaterDrop
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.intellidream.daily.model.HabitDrinkBreakdown
import com.intellidream.daily.model.HabitType
import kotlin.math.roundToInt

data class AggregatedTickerItem(
    val id: String,
    val name: String,
    val amount: Double,
    val unit: String,
    val color: Color,
    val icon: ImageVector
)

@Composable
fun HabitCircleBreakdownTicker(
    items: List<HabitDrinkBreakdown>,
    habitType: HabitType = HabitType.WATER,
    modifier: Modifier = Modifier
) {
    val aggregatedItems = remember(items, habitType) {
        aggregateItems(items, habitType)
    }

    if (aggregatedItems.isEmpty()) return

    val scrollState = rememberScrollState()

    Box(
        modifier = modifier
            .widthIn(max = 175.dp)
            .height(24.dp)
            .shadow(4.dp, CircleShape)
            .clip(CircleShape)
            .background(Color.Black.copy(alpha = 0.40f))
            .border(0.8.dp, Color.White.copy(alpha = 0.18f), CircleShape)
            .padding(horizontal = 6.dp, vertical = 2.dp),
        contentAlignment = Alignment.Center
    ) {
        Row(
            modifier = Modifier.horizontalScroll(scrollState),
            horizontalArrangement = Arrangement.spacedBy(5.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            aggregatedItems.forEach { item ->
                val badgeText = formatBadgeText(item, habitType, aggregatedItems.size == 1)

                Row(
                    modifier = Modifier
                        .clip(CircleShape)
                        .background(item.color.copy(alpha = 0.22f))
                        .border(0.8.dp, item.color.copy(alpha = 0.40f), CircleShape)
                        .padding(horizontal = 6.dp, vertical = 2.dp),
                    horizontalArrangement = Arrangement.spacedBy(3.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Icon(
                        imageVector = item.icon,
                        contentDescription = null,
                        tint = item.color,
                        modifier = Modifier.size(9.dp)
                    )
                    Text(
                        text = badgeText,
                        color = Color.White.copy(alpha = 0.92f),
                        fontSize = 9.5.sp,
                        fontWeight = FontWeight.Bold
                    )
                }
            }
        }
    }
}

private fun formatBadgeText(item: AggregatedTickerItem, habitType: HabitType, isSingle: Boolean): String {
    val amountInt = item.amount.roundToInt()
    val formattedAmount = if (item.unit == "ml") "$amountInt ml" else "$amountInt"

    if (isSingle) {
        return "${item.name}: $formattedAmount"
    }

    return if (habitType == HabitType.WATER) {
        if (item.name == "Water") formattedAmount else "${item.name} $formattedAmount"
    } else {
        when (item.name) {
            "Cigarette" -> "$amountInt Cig"
            "Heated" -> "$amountInt Heat"
            "Rolled" -> "$amountInt Roll"
            "Cigarillo" -> "$amountInt Cigar"
            else -> "$amountInt ${item.name}"
        }
    }
}

private fun aggregateItems(
    items: List<HabitDrinkBreakdown>,
    habitType: HabitType
): List<AggregatedTickerItem> {
    val map = mutableMapOf<String, AggregatedTickerItem>()

    for (item in items) {
        val (category, color, icon) = if (habitType == HabitType.WATER) {
            normalizeDrink(item.drink, item.hexColor)
        } else {
            normalizeSmoke(item.drink, item.hexColor)
        }

        val existing = map[category]
        if (existing != null) {
            map[category] = existing.copy(amount = existing.amount + item.amount)
        } else {
            map[category] = AggregatedTickerItem(
                id = category,
                name = category,
                amount = item.amount,
                unit = item.unit,
                color = color,
                icon = icon
            )
        }
    }

    return map.values.sortedByDescending { it.amount }
}

private fun normalizeDrink(name: String, defaultColorHex: String): Triple<String, Color, ImageVector> {
    val lower = name.lowercase()
    return when {
        lower.contains("coffee") || lower.contains("espresso") || lower.contains("cappuccino") || lower.contains("latte") ->
            Triple("Coffee", Color(0xFFF59E0B), Icons.Rounded.Coffee)
        lower.contains("tea") || lower.contains("matcha") || lower.contains("infusion") ->
            Triple("Tea", Color(0xFF84CC16), Icons.Rounded.EmojiFoodBeverage)
        lower.contains("water") || lower.contains("glass") || lower.contains("bottle") ->
            Triple("Water", Color(0xFF00E5FF), Icons.Rounded.WaterDrop)
        else ->
            Triple(name, parseColorSafely(defaultColorHex, Color(0xFF00E5FF)), Icons.Rounded.WaterDrop)
    }
}

private fun normalizeSmoke(name: String, defaultColorHex: String): Triple<String, Color, ImageVector> {
    val lower = name.lowercase()
    return when {
        lower.contains("cigarette") || lower == "cig" || lower.contains("standard") ->
            Triple("Cigarette", Color(0xFFEF4444), Icons.Rounded.LocalFireDepartment)
        lower.contains("cigarillo") || (lower.contains("cigar") && !lower.contains("cigarette")) ->
            Triple("Cigarillo", Color(0xFFA855F7), Icons.Rounded.SmokingRooms)
        lower.contains("heat") || lower.contains("iqos") || lower.contains("glo") || lower.contains("vape") ->
            Triple("Heated", Color(0xFF3B82F6), Icons.Rounded.Bolt)
        lower.contains("roll") ->
            Triple("Rolled", Color(0xFFF97316), Icons.Rounded.Eco)
        else ->
            Triple("Cigarette", parseColorSafely(defaultColorHex, Color(0xFFEF4444)), Icons.Rounded.LocalFireDepartment)
    }
}

private fun parseColorSafely(hex: String, defaultColor: Color): Color {
    return try {
        val cleanHex = hex.removePrefix("#")
        val colorInt = cleanHex.toLong(16)
        if (cleanHex.length == 6) {
            Color(0xFF000000 or colorInt)
        } else if (cleanHex.length == 8) {
            Color(colorInt)
        } else {
            defaultColor
        }
    } catch (_: Exception) {
        defaultColor
    }
}
